### PR TITLE
caf: buttons: Fix buttons hang on consumed wake_up_event

### PR DESCRIPTION
**Pull request by emob-nordic**
Fix issue when buttons module can hang if wake_up_event
submitted on button press in STATE_IDLE mode is consumed by
power manager.

This change explicitly adds missing STATE_RESUMING when
button is pressed in STATE_IDLE, interrupts are turned off
and module is expecting wake_up_event.

If the wake_up_event does not come back after timeout
specified by CONFIG_CAF_BUTTONS_PM_WAKEUP_TIMEOUT timeout,
the interrupts for buttons are re-enabled.

Documentation updated accordingly.

Jira: NCSDK-13619

### CODE REVIEW

File: doc/nrf/releases/release-notes-changelog.rst
Comment (divipillai) [inline doc/nrf/releases/release-notes-changelog.rst:247]:
```suggestion
  * Fixed the NCSDK-13619 issue where :ref: `caf_buttons` module can hang in idle state if c:struct:`wake_up_event` is called on button press.
```

Comment (emob-nordic) [reply doc/nrf/releases/release-notes-changelog.rst:247]: I need to add an info that the event is latar consumed. Could it be:
```
* Fixed the NCSDK-13619 issue where :ref:`caf_buttons` module can hang in idle state if :c:struct:`wake_up_event` is submitted on button press and later consumed by any other module.
```
?

Comment (divipillai) [reply doc/nrf/releases/release-notes-changelog.rst]: yes, this looks fine.

**Review by divipillai — APPROVED** [review]

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: subsys/caf/modules/buttons.c
Comment (emob-nordic) [reply subsys/caf/modules/buttons.c:375]: That is a mistake, thanks.

File: subsys/caf/modules/buttons.c
Comment (emob-nordic) [reply subsys/caf/modules/buttons.c:364]: It should, removed.

File: subsys/caf/modules/Kconfig.buttons
Comment (emob-nordic) [reply subsys/caf/modules/Kconfig.buttons:37]: Will remove it.

File: subsys/caf/modules/Kconfig.buttons
Comment (emob-nordic) [reply subsys/caf/modules/Kconfig.buttons:35]: `Module interrupt re-enable timeout in ms` - same as for other timeouts in this module.

File: doc/nrf/libraries/caf/buttons.rst
Comment (emob-nordic) [reply doc/nrf/libraries/caf/buttons.rst:140]: My bad, i prepared incorrect state machine drawing and based on that this was added. I will remove it and correct the diagram too.

**Comment by divipillai** [comment]
Added @Avalei for image review.

**Review by divipillai — APPROVED** [review]

**Review by Avalei — CHANGES_REQUESTED** [review]
For the caf_button_states.svg we need to make a couple small changes:

- For [state diagrams](https://projecttools.nordicsemi.no/confluence/display/TECHDOC/State+diagram) we are now using curved lines instead of straight or angled ones. To edit this, right-click on the line and choose the "Curved Connector" option.
- The power_down_event text on the left side should be rotated 180 degrees. We want all of the text readable from top to down in the same direction.
- For the power_down_event on the left side, can we move it out to the right more so the underscores are more visible? It blends in with the line right now.
- The diagonal power_down_event in the upper left area looks a little bit crooked compared to the line. Is is possible to align them a little more? 
- Can you expand the text bounding box for "Initial scan" so that the text is all on one line?

**Review by Avalei — APPROVED** [review]
Figure is approved.

**Comment by MarekPieta** [comment]
Please do not merge it until we will have review from the code owner.

**Comment by emob-nordic** [comment]
Updated for 2.0.0 milestone. We don't need this fix in 1.9.0 release as it doesn't break anything for 1.9.0. Sorry for the noise.

File: subsys/caf/modules/Kconfig.buttons
Comment (MarekPieta) [inline subsys/caf/modules/Kconfig.buttons:40]: I would not say `by power manager` to avoid the reference here. If any other module consumes it, the outcome would be the same.

File: doc/nrf/libraries/caf/buttons.rst
Comment (MarekPieta) [inline doc/nrf/libraries/caf/buttons.rst:136]: :c:enumerator:`STATE_SCANNING`? Also for other states. The state are values of enum.
That would be also consistent with BLE bond doc.

File: subsys/caf/modules/buttons.c
Comment (MarekPieta) [inline subsys/caf/modules/buttons.c:207]: Add brackets, please

File: doc/nrf/libraries/caf/images/caf_buttons_states.svg
Comment (MarekPieta) [inline doc/nrf/libraries/caf/images/caf_buttons_states.svg:5]: On diagram the arrow on button press between state resuming and state idle is inversed. Then we go from idle to resuming state.

**Review by MarekPieta — APPROVED** [review]

File: doc/nrf/libraries/caf/buttons.rst
Comment (emob-nordic) [reply doc/nrf/libraries/caf/buttons.rst:136]: okay.

File: doc/nrf/libraries/caf/images/caf_buttons_states.svg
Comment (emob-nordic) [reply doc/nrf/libraries/caf/images/caf_buttons_states.svg:5]: @divipillai we need to fix that, i didnt spot it by myself. Please fetch before you add changes as i fixed other stuff that Marek spotted.

File: subsys/caf/modules/buttons.c
Comment (rakons) [inline subsys/caf/modules/buttons.c:30]: I really do not like if CONFIG_ setting are overwritten this way.
Use rather:
```
#ifdef CONFIG_CAF_BUTTONS_PM_WAKEUP_TIMEOUT
#define BUTTONS_PM_WAKEUP_TIMEOUT CONFIG_CAF_BUTTONS_PM_WAKEUP_TIMEOUT
#else
#define BUTTONS_PM_WAKEUP_TIMEOUT 0
#endif
```
And then use BUTTONS_PM_WAKEUP_TIMEOUT in the code.

**Review by rakons — APPROVED** [review]

**Comment by carlescufi** [comment]
@emob-nordic why is this DNM?

**Comment by emob-nordic** [comment]
> @emob-nordic why is this DNM?

I was marked DNM during the ncs 1.9.0 release and not touched since then. Let's keep this DNM until i find some time to check if its safe to merge. I will ask you to merge then, is that okay?

**Comment by thst-nordic** [comment]
can you force push?   this will clear the old check items.  
image

**Comment by mbolivar-nordic** [comment]
> can you force push? this will clear the old check items.

@emob-nordic please handle this or remove the 2.0.0 milestone if this PR is not needed for this release

**Comment by gte1** [comment]
Was this supposed to be reconsidered? 

> Hi @emob-nordic , let's reconsider this after the release.

**Comment by emob-nordic** [comment]
> Was this supposed to be reconsidered?
> 
> > Hi @emob-nordic , let's reconsider this after the release.

We are not proceeding this as a priority now.

**Pull request closed without merging**
